src/peergos/server/space/JdbcUsageStore.java

```diff
@@ -0,0 +1,370 @@
+        try (PreparedStatement insert = conn.prepareStatement("UPDATE userusage SET errored = ? " +
+                "WHERE user_id = ?;")) {
+            insert.setBoolean(1, errored);
+            insert.setInt(3, userId);
+            boolean updated1Record = insert.execute();
+            if (! updated1Record)
+                throw new IllegalStateException("Didn't update one record!");
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    @Override
+    public UserUsage getUsage(String username) {
+        int userId = getUserId(username);
+        try (PreparedStatement search = conn.prepareStatement("SELECT total_bytes, errored FROM userusage WHERE user_id = ?;")) {
+            search.setInt(1, userId);
+            ResultSet resultSet = search.executeQuery();
+            resultSet.next();
+            long totalBytes = resultSet.getLong(1);
+            boolean errored = resultSet.getBoolean(2);
+            Map<PublicKeyHash, Long> pending = new HashMap<>();
+            try (PreparedStatement pendingSearch = conn.prepareStatement("SELECT writer_id, pending_bytes FROM pendingusage WHERE user_id = ?;")) {
+                pendingSearch.setInt(1, userId);
+                ResultSet pendingResults = pendingSearch.executeQuery();
+                while (pendingResults.next())
+                    pending.put(getWriter(pendingResults.getInt(1)), pendingResults.getLong(2));
+            } catch (SQLException sqe) {
+                LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+                throw new RuntimeException(sqe);
+            }
+            return new UserUsage(totalBytes, errored, pending);
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    private int getUserId(String username) {
+        try (PreparedStatement insert = conn.prepareStatement("SELECT id FROM users WHERE name = ?;")) {
+            insert.setString(1, username);
+            ResultSet resultSet = insert.executeQuery();
+            resultSet.next();
+            return resultSet.getInt(1);
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    private int getWriterId(PublicKeyHash writer) {
+        try (PreparedStatement insert = conn.prepareStatement("SELECT id FROM writers WHERE key_hash = ?;")) {
+            insert.setBytes(1, writer.toBytes());
+            ResultSet resultSet = insert.executeQuery();
+            resultSet.next();
+            return resultSet.getInt(1);
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    @Override
+    public void addWriter(String owner, PublicKeyHash writer) {
```

Comment:
the sql statments should be in a db transaction

Reply:
In this case I don't think they need a transaction